src/Language/PureScript/Linter.hs

```diff
@@ -75,17 +78,42 @@ lint (Module _ _ mn ds _) = censor (onErrorMessages (ErrorInModule mn)) $ mapM_
     stepD s _ = (s, mempty)
 
     stepE :: S.Set Ident -> Expr -> (S.Set Ident, MultipleErrors)
-    stepE s (Abs (Left name) _) = bind s name
+    stepE s (Abs (Left name) _) = bindName s name
     stepE s (Let ds' _) =
-      case mapAccumL bind s (nub (mapMaybe getDeclIdent ds')) of
+      case mapAccumL bindName s (nub (mapMaybe getDeclIdent ds')) of
         (s', es) -> (s', mconcat es)
     stepE s _ = (s, mempty)
 
     stepB :: S.Set Ident -> Binder -> (S.Set Ident, MultipleErrors)
-    stepB s (VarBinder name) = bind s name
-    stepB s (NamedBinder name _) = bind s name
+    stepB s (VarBinder name) = bindName s name
+    stepB s (NamedBinder name _) = bindName s name
     stepB s _ = (s, mempty)
 
-    bind :: S.Set Ident -> Ident -> (S.Set Ident, MultipleErrors)
-    bind s name | name `S.member` s = (s, errorMessage (ShadowedName name))
-                | otherwise = (S.insert name s, mempty)
+    bindName :: S.Set Ident -> Ident -> (S.Set Ident, MultipleErrors)
+    bindName = bind ShadowedName
+
+  checkTypeVarsInDecl :: Declaration -> MultipleErrors
+  checkTypeVarsInDecl d = let (f, _, _, _, _) = accumTypes checkTypeVars in f d
+
+  checkTypeVars :: Type -> MultipleErrors
+  checkTypeVars ty = everythingWithContextOnTypes S.empty mempty mappend step ty <> findUnused ty
+    where
+    step :: S.Set String -> Type -> (S.Set String, MultipleErrors)
+    step s (ForAll tv _ _) = bindVar s tv
+    step s _ = (s, mempty)
+    bindVar :: S.Set String -> String -> (S.Set String, MultipleErrors)
+    bindVar = bind ShadowedTypeVar
+    findUnused :: Type -> MultipleErrors
```

Comment:
Oh, actually, I'm not sure it is - we don't get the accumulated context back when using `everythingWithContextOnTypes`, unless you had something else in mind?